File: two pointer/727. Minimum Window Subsequence/py/main.py
import math

# Time O(N*M)
# Space O(1)
class Solution:
    def minWindow(self, S, T):
      end = 0
      minIndex = math.inf
      minLen = math.inf

      while end < len(S):
        tIndex = 0
        while end < len(S) and tIndex < len(T):
          if S[end] == T[tIndex]:
            tIndex += 1
          if tIndex == len(T):
            break
          end += 1

        start = end
        tIndex = len(T) - 1
        
        if end == len(S):
            break
        
        while start >= 0:
          if S[start] == T[tIndex]:
            tIndex -= 1

          if tIndex < 0:
            break
          start -= 1

        if minLen > end - start + 1:
          minLen = end - start + 1
          minIndex = start

        end = start + 1

      return '' if minLen == math.inf else S[minIndex:minIndex + minLen]

File: two pointer/727. Minimum Window Subsequence/js/main.js
// Time O(N*M)
// Space O(1)
const minWindow = (S, T) => {
  let endIndex = 0;
  let minLen = Number.MAX_VALUE;
  let minStartIndex = 0;

  while (endIndex < S.length) {
    let tIndex = 0;

    while (endIndex < S.length && tIndex < T.length) {
      if (S[endIndex] == T[tIndex]) {
        tIndex++;
      }

      if (tIndex == T.length) {
        break;
      }

      endIndex++;
    }

    if (endIndex == S.length) {
      break;
    }

    let startIndex = endIndex;
    tIndex = T.length - 1;

    while (startIndex >= 0) {
      if (S[startIndex] == T[tIndex]) {
        tIndex--;
      }

      if (tIndex < 0) {
        break;
      }

      startIndex--;
    }

    if (endIndex - startIndex + 1 < minLen) {
      minLen = endIndex - startIndex + 1;
      minStartIndex = startIndex;
    }

    endIndex = startIndex + 1;
  }

  return minLen === Number.MAX_VALUE ? '' : S.substring(minStartIndex, minStartIndex + minLen);
};
